### PR TITLE
chainntnfs: Isolate conf notifications during historical scans

### PR DESCRIPTION
**Pull request by cfromknecht**
In this PR, we modify the `TxConfNotifier` to isolate confirmation notifications while a historical dispatch is being processed. The purpose of this is to segment which notifications should have their height hints updated in the cache. During a historical rescan, we don't yet know if the txn is in the chain, so we shouldn't continue to update its height hint until after finishing the rescan. Once the rescan finishes, the notification is grouped along with all other notifications at tip so that their height hints can progress as blocks are added.

This PR also introduces:
 - a reorg safety delta to height hints, currently 144 blocks
   - with the recent LRU cache added to neutrino, this shouldn't result too much unnecessary network I/O
   - the height hint is still clamped to whatever the client initial prescribes, so initial registration doesn't take a hit either
 - moves querying of confirmation height hints into the `TxConfNotifier`
 - reorders the persistence/spend-notification dispatch to improve external consistency

In a follow PR, we will do the same for spend notifications.

This PR depends on #1951, which contains the first four commits

### CODE REVIEW

**Comment by halseth** [comment]
Is this needed for initial height hint cache enabling?

**Comment by cfromknecht** [comment]
@halseth yes, and similar work needs to be done for spends

File: chainntnfs/txconfnotifier.go [hunk 1]
```diff
@@ -101,6 +101,17 @@ type TxConfNotifier struct {
 	sync.Mutex
 }
 
+type confNtfnSet struct {
+	ntfns       map[uint64]*ConfNtfn
+	rescanCount int
+}
+
+func newConfNtfnSet() *confNtfnSet {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: these should have godocs.

File: chainntnfs/txconfnotifier.go [hunk 2]
```diff
@@ -124,35 +135,35 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32,
...
-		)
-		if err != nil {
-			// The error is not fatal, so we should not return an
-			// error to the caller.
-			Log.Errorf("Unable to update confirm hint to %d for "+
-				"%v: %v", tcn.currentHeight, *ntfn.TxID, err)
-		}
+	confSet, ok := tcn.confNotifications[*ntfn.TxID]
+	if !ok {
+		confSet = newConfNtfnSet()
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: Some comments here would be appreciated.

File: chainntnfs/txconfnotifier.go [hunk 3]
```diff
@@ -142,6 +147,17 @@ func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn) (bool, uint32, error) {
 	default:
 	}
 
+	// Before proceeding to register the notification, we'll query our
+	// height hint cache to determine whether a better one exists.
+	hint, err := tcn.hintCache.QueryConfirmHint(*ntfn.TxID)
+	if err == nil {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: log error?

File: chainntnfs/txconfnotifier.go [hunk 4]
```diff
@@ -124,35 +140,46 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32,
...
+	if !ok {
+		confSet = newConfNtfnSet()
+		tcn.confNotifications[*ntfn.TxID] = confSet
 	}
 
-	ntfns[ntfn.ConfID] = ntfn
+	confSet.ntfns[ntfn.ConfID] = ntfn
+
+	switch {
+	case ntfn.HeightHint > tcn.currentHeight:
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: Also comments on what is happening here.

File: chainntnfs/txconfnotifier.go [hunk 5]
```diff
@@ -177,91 +188,106 @@ func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
...
+	}
 
-		// We'll send a 0 value to the Updates channel, indicating that
-		// the transaction has already been confirmed.
-		select {
-		case ntfn.Event.Updates <- 0:
-		case <-tcn.quit:
-			return ErrTxConfNotifierExiting
+	// Update the conf details of all ntfns that don't yet have them.
+	for _, ntfn := range confSet.ntfns {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: Isn't this already handled by each rescan calling `UpdateConfDetails` for its client ID?

File: chainntnfs/txconfnotifier.go [hunk 6]
```diff
@@ -335,7 +366,11 @@ func (tcn *TxConfNotifier) ConnectTip(blockHash *chainhash.Hash,
 		txsToUpdateHints = append(txsToUpdateHints, confirmedTx)
 	}
 out:
-	for maybeUnconfirmedTx := range tcn.confNotifications {
+	for maybeUnconfirmedTx, confSet := range tcn.confNotifications {
+		if confSet.rescanCount > 0 {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: add comment

File: chainntnfs/txconfnotifier.go [hunk 7]
```diff
@@ -470,7 +506,8 @@ func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
 	// clients is always non-blocking.
 	for initialHeight, txHashes := range tcn.txsByInitialHeight {
 		for txHash := range txHashes {
-			for _, ntfn := range tcn.confNotifications[txHash] {
+			confSet := tcn.confNotifications[txHash]
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: any reason `!ok` check is skipped here?

File: chainntnfs/txconfnotifier.go (hunk 1 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 5 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: only one rescan is dispatched now

File: chainntnfs/txconfnotifier.go (hunk 7 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: the assumption is that the keys in `txsByInitialHeight` are a subset of `confNotifications`, though it doesn't hurt to add the check

File: chainntnfs/txconfnotifier_test.go [hunk 8]
```diff
@@ -715,65 +720,105 @@ func TestTxConfHeightHintCache(t *testing.T) {
...
+	_, err = hintCache.QueryConfirmHint(tx2Hash)
+	if err != chainntnfs.ErrConfirmHintNotFound {
+		t.Fatalf("unexpected error when querying for height hint "+
+			"want: %v, got %v",
+			chainntnfs.ErrConfirmHintNotFound, err)
+	}
+
+	// Create a new block that will include the dummy transaction and extend
+	// the chain.
+	txDummy := wire.MsgTx{Version: 3}
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier_test.go]: just needs a unique txid

**Comment by halseth** [comment]
Needs a rebase! (I think?)

File: chainntnfs/txconfnotifier.go [hunk 9]
```diff
@@ -225,14 +225,37 @@ func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn) (bool, uint32, error) {
...
+	}
+
+	// If the provided or cached height hint indicates that the transaction
+	// is to be confirmed at a height greater than the conf notifier's
+	// current height, we'll refrain from spawning a historical dispatch.
+	if startHeight > tcn.currentHeight {
+		// Set the rescan status to complete, which will allow the conf
+		// notifier to start delivering messages for this set
+		// immediately.
+		confSet.rescanStatus = rescanComplete
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: Yes I believe that's correct, only the height hint of the first registration would be considered in that case. The alternative is to do some ARC magic and actually maintain a count of all active rescans, at the expense of (probably) rescanning unnecessarily in most instances. This was the approach taken early on, though I agree it'd be better in the long run to just ensure we are passing them in properly instead of being wasteful in manual rescans.

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 10]
```diff
@@ -277,14 +277,12 @@ out:
...
 
-					if confDetails != nil {
-						err := b.txConfNotifier.UpdateConfDetails(
-							*msg.TxID, msg.ConfID,
-							confDetails,
-						)
-						if err != nil {
-							chainntnfs.Log.Error(err)
-						}
+					err = b.txConfNotifier.UpdateConfDetails(
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/bitcoindnotify/bitcoind.go]: fixed

File: chainntnfs/txconfnotifier.go [hunk 11]
```diff
@@ -177,91 +188,106 @@ func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
...
 	// this transaction with the given ID.
-	ntfns, ok := tcn.confNotifications[txid]
+	confSet, ok := tcn.confNotifications[txid]
 	if !ok {
-		return fmt.Errorf("no notifications found for txid %v", txid)
+		return fmt.Errorf("no notification found with TxID %v", txid)
 	}
 
-	ntfn, ok := ntfns[clientID]
+	ntfn, ok := confSet.ntfns[clientID]
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: I don't think so, otherwise we will dereference a nil confSet

[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: ahh nvm, misread. this is done in a later commit

File: chainntnfs/txconfnotifier.go [hunk 12]
```diff
@@ -299,7 +325,12 @@ func (tcn *TxConfNotifier) ConnectTip(blockHash *chainhash.Hash,
 	// handled correctly.
 	for _, tx := range txns {
 		txHash := tx.Hash()
-		for _, ntfn := range tcn.confNotifications[*txHash] {
+		confSet, ok := tcn.confNotifications[*txHash]
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: lnd_test.go [hunk 13]
```diff
@@ -7532,13 +7532,7 @@ func testGraphTopologyNotifications(net *lntest.NetworkHarness, t *harnessTest)
...
 		case graphUpdate := <-graphSub.updateChan:
 
-			if len(graphUpdate.ChannelUpdates) > 0 {
-				chanUpdate := graphUpdate.ChannelUpdates[0]
-				if chanUpdate.Capacity != int64(chanAmt) {
-					t.Fatalf("channel capacities mismatch:"+
-						" expected %v, got %v", chanAmt,
-						btcutil.Amount(chanUpdate.Capacity))
-				}
+			for _, chanUpdate := range graphUpdate.ChannelUpdates {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply lnd_test.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: it technically would be possible for the confset to track the height hints it's seen so that we can be alerted of the situation, but maybe something to leave to a later PR

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 14]
```diff
@@ -956,12 +947,20 @@ func (b *BitcoindNotifier) RegisterConfirmationsNtfn(txid *chainhash.Hash,
 		HeightHint:       heightHint,
 	}
 
-	if err := b.txConfNotifier.Register(ntfn); err != nil {
+	chainntnfs.Log.Infof("New confirmation subscription: "+
+		"txid=%v, numconfs=%v", txid, numConfs)
+
+	dispatch, err := b.txConfNotifier.Register(ntfn)
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/bitcoindnotify/bitcoind.go]: done!

File: chainntnfs/txconfnotifier.go [hunk 15]
```diff
@@ -177,91 +188,106 @@ func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
...
 	}
 
-	ntfn, ok := ntfns[clientID]
+	ntfn, ok := confSet.ntfns[clientID]
 	if !ok {
 		return fmt.Errorf("no notification found with ID %v", clientID)
 	}
 
+	// Decrement the rescan count for each registered client that finishes.
+	confSet.rescanCount--
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: since this is removed in a later commit, they should be squashed together to not having to introduce it in the first place.

File: chainntnfs/txconfnotifier.go [hunk 16]
```diff
@@ -150,14 +150,23 @@ const (
 	rescanComplete
 )
 
+// confNtfnSet holds all known, registered confirmation notifications for a
+// single txid. If duplicates notifications are requested, only one historical
+// dispatch will be spawned to ensure redundant scans are not permitted. A
+// single conf detail will be constructed and dispatched to all interested
+// clients.
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: Squash with the commit introducing `confNtfnSet` :)

File: chainntnfs/txconfnotifier.go [hunk 17]
```diff
@@ -242,6 +239,10 @@ func (tcn *TxConfNotifier) Register(
 	// A prior rescan has already completed and we are actively watching at
 	// tip for this txid.
 	case rescanComplete:
+		// If conf details for this set of notifications has already
+		// been found, we'll attempt to deliver them immediately to this
+		// client.
+		tcn.dispatchConfDetails(ntfn, confSet.details)
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: also squash?

File: chainntnfs/txconfnotifier_test.go [hunk 18]
```diff
@@ -124,7 +124,7 @@ func TestTxConfFutureDispatch(t *testing.T) {
 		NumConfirmations: tx1NumConfs,
 		Event:            chainntnfs.NewConfirmationEvent(tx1NumConfs),
 	}
-	if _, _, err := tcn.Register(&ntfn1); err != nil {
+	if _, err := tcn.Register(&ntfn1); err != nil {
```
Comment (halseth) [inline chainntnfs/txconfnotifier_test.go]: looks like also can just be squashed with the prev commit changing this

File: chainntnfs/txconfnotifier.go (hunk 16 above)
Comment (halseth) [reply chainntnfs/txconfnotifier.go]: Ok, several more "fixup" commits that are not marked fixup so will not comment on them for now :)

File: chainntnfs/txconfnotifier_test.go (hunk 18 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier_test.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 17 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 15 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go [hunk 19]
```diff
@@ -320,10 +333,20 @@ func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
...
 	// ConnectTip.
-	if details == nil || details.BlockHeight > tcn.currentHeight {
+	if details == nil {
+		Log.Debugf("Conf details for txid=%v not found during "+
+			"historical dispatch, waiting to dispatch at tip", txid)
+		return nil
+	}
+
+	if details.BlockHeight > tcn.currentHeight {
+		Log.Debugf("Conf details for txid=%v found above current "+
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: I think it can be during a reorg

File: chainntnfs/txconfnotifier.go [hunk 20]
```diff
@@ -48,6 +48,29 @@ type ConfNtfn struct {
...
+// transaction identifier. The parameters include the start and end block
+// heights specifying the range of blocks to scan.
+type HistoricalConfDispatch struct {
+	// TxID is the transaction ID to search for in the historical dispatch.
+	TxID *chainhash.Hash
+
+	// PkScript is a public key script from an output created by this
+	// transaction.
+	//
+	// NOTE: This value MUST be set when the dispatch is to be performed
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: It doesn't necessarily need to be for txindex backends

File: chainntnfs/txconfnotifier.go [hunk 21]
```diff
@@ -331,72 +332,96 @@ func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash,
...
+	// If no details are provided, return early as we can't dispatch.
+	if details == nil {
+		return nil
+	}
+
+	// Set the confirmation details for this notification, only if the
+	// notification doesn't already have details. This ensure we only fall
+	// through the following logic at most once, which could cause the
+	// buffered channels to block when exceeding their allocated capacity.
+	if ntfn.details != nil {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: I think this could have been hit when we dispatched multiple historical rescans. I feel like it's a useful sanity check though.

[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: reworked anyways to use confset details

File: chainntnfs/txconfnotifier.go [hunk 22]
```diff
@@ -459,17 +459,26 @@ func (tcn *TxConfNotifier) ConnectTip(blockHash *chainhash.Hash,
...
+		Log.Debugf("Block contains txid=%v, constructing details",
+			txHash)
+
+		details := &TxConfirmation{
+			BlockHash:   blockHash,
+			BlockHeight: blockHeight,
+			TxIndex:     uint32(tx.Index()),
+		}
+
+		confSet.details = details
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go [hunk 23]
```diff
@@ -548,7 +579,14 @@ out:
 	// this new height.
 	for _, txHashes := range tcn.txsByInitialHeight {
 		for txHash := range txHashes {
-			confSet := tcn.confNotifications[txHash]
+			confSet, ok := tcn.confNotifications[txHash]
+			if !ok {
+				Log.Warnf("Txns by initial height index "+
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: Don't know if it's worth panicking in production

File: chainntnfs/txconfnotifier.go [hunk 24]
```diff
@@ -101,6 +135,47 @@ type TxConfNotifier struct {
...
+
+// confNtfnSet holds all known, registered confirmation notifications for a
+// single txid. If duplicates notifications are requested, only one historical
+// dispatch will be spawned to ensure redundant scans are not permitted. A
+// single conf detail will be constructed and dispatched to all interested
+// clients.
+type confNtfnSet struct {
+	ntfns        map[uint64]*ConfNtfn
+	rescanStatus rescanState
+	details      *TxConfirmation
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go [hunk 25]
```diff
@@ -791,6 +749,48 @@ func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
...
+	// dispatched, we'll need to clear its entry within the
+	// ntfnsByConfirmHeight index to prevent from notifiying the client once
+	// the notifier reaches the confirmation height.
+	if !ntfn.dispatched {
+		confHeight := heightDisconnected + ntfn.NumConfirmations - 1
+		ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
+		if exists {
+			delete(ntfnSet, ntfn)
+		}
+	} else {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: code catch, i think rebasing moved some code around. fixed

File: chainntnfs/txconfnotifier.go [hunk 26]
```diff
@@ -335,7 +551,14 @@ func (tcn *TxConfNotifier) ConnectTip(blockHash *chainhash.Hash,
 		txsToUpdateHints = append(txsToUpdateHints, confirmedTx)
 	}
 out:
-	for maybeUnconfirmedTx := range tcn.confNotifications {
+	for maybeUnconfirmedTx, confSet := range tcn.confNotifications {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: This logic is a bit hard to follow currently. Would it be equivalent to do
```golang
var txsToUpdateHints []chainhash.Hash
for maybeUnconfirmedTx, confSet := range tcn.confNotifications {
                if confSet.rescanStatus != rescanComplete {
			continue
		}
                txsToUpdateHints = append(txsToUpdateHints, maybeUnconfirmedTx)
}
```
?

File: chainntnfs/txconfnotifier.go [hunk 27]
```diff
@@ -370,16 +593,24 @@ out:
...
+				Log.Warnf("Txns by initial height index "+
+					"contains key not known to "+
+					"confNotifications: %v", txHash)
+				continue
+			}
+
+			// If the confset hasn't learned about the confirmation
+			// of its transaction yet, we can't schedule the
+			// dispatch.
+			if confSet.details == nil {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: I think this is also an error?

File: chainntnfs/txconfnotifier.go [hunk 28]
```diff
@@ -403,12 +634,25 @@ out:
...
-	for ntfn := range tcn.ntfnsByConfirmHeight[tcn.currentHeight] {
+	for ntfn := range tcn.ntfnsByConfirmHeight[blockHeight] {
+		confSet, ok := tcn.confNotifications[*ntfn.TxID]
+		if !ok {
+			Log.Warnf("Conf set not found for scheduled dispatch "+
+				"of txid=%x", ntfn.TxID)
+			continue
+		}
+
+		if confSet.details == nil {
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: I would actually prefer just to use the details directly without checking for `nil` here and other places where it should be set. If that happens it is a quite serious bug and will be buried in the logs if we just print and continue.

File: chainntnfs/txconfnotifier.go [hunk 29]
```diff
@@ -535,6 +757,49 @@ func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
...
+	// If the transaction's confirmation notification has yet to be
+	// dispatched, we'll need to clear its entry within the
+	// ntfnsByConfirmHeight index to prevent from notifiying the client once
+	// the notifier reaches the confirmation height.
+	if !ntfn.dispatched {
+		confHeight := heightDisconnected + ntfn.NumConfirmations - 1
+		ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
+		if exists {
+			delete(ntfnSet, ntfn)
+		}
```
Comment (halseth) [inline chainntnfs/txconfnotifier.go]: just return here to avoid indentation, or change the order back to the original.

File: chainntnfs/btcdnotify/btcd.go [hunk 30]
```diff
@@ -324,38 +324,34 @@ out:
...
-					msg.TxID, msg.NumConfirmations)
-
-				bestHeight := uint32(b.bestBlock.Height)
-
-				// Look up whether the transaction is already
-				// included in the active chain. We'll do this
-				// in a goroutine to prevent blocking
-				// potentially long rescans.
+			case *chainntnfs.HistoricalConfDispatch:
+				// Look up whether the transaction is already included in the active
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd.go]: long line

File: chainntnfs/neutrinonotify/neutrino.go [hunk 31]
```diff
@@ -314,28 +314,17 @@ out:
...
-				n.heightMtx.RLock()
-				currentHeight := n.bestHeight
-				n.heightMtx.RUnlock()
-
-				// Look up whether the transaction is already
-				// included in the active chain. We'll do this
-				// in a goroutine to prevent blocking
-				// potentially long rescans.
+			case *chainntnfs.HistoricalConfDispatch:
+				// Look up whether the transaction is already included in the active
```
Comment (halseth) [inline chainntnfs/neutrinonotify/neutrino.go]: same

File: chainntnfs/neutrinonotify/neutrino.go [hunk 32]
```diff
@@ -347,20 +336,27 @@ out:
...
+					// rescans have now completed.
+					err = n.txConfNotifier.UpdateConfDetails(
+						*msg.TxID, confDetails,
 					)
 					if err != nil {
 						chainntnfs.Log.Error(err)
 					}
 
 					if confDetails != nil {
-						err := n.txConfNotifier.UpdateConfDetails(
```
Comment (halseth) [inline chainntnfs/neutrinonotify/neutrino.go]: with this change, can the below code be removed?

File: chainntnfs/txconfnotifier.go [hunk 33]
```diff
@@ -119,40 +194,113 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32,
...
 	select {
 	case <-tcn.quit:
-		return ErrTxConfNotifierExiting
+		return nil, ErrTxConfNotifierExiting
 	default:
 	}
 
+	// Enforce that we will not dispatch confirmations beyond the reorg
+	// safety limit.
+	if ntfn.NumConfirmations > tcn.reorgSafetyLimit {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: the txconfnotifier will remove certain indexes it maintains after the reorg height, so we can't reliably claim to deliver ntfns if the caller requests more than the reorg limit

File: chainntnfs/neutrinonotify/neutrino.go (hunk 32 above)
Comment (cfromknecht) [reply chainntnfs/neutrinonotify/neutrino.go]: wdym?

File: chainntnfs/txconfnotifier.go (hunk 26 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed wilmer's branch https://github.com/lightningnetwork/lnd/pull/2004/commits/e47bbaff3a6d1fc0f39f361c01557fa986d3653f

File: chainntnfs/btcdnotify/btcd.go (hunk 30 above)
Comment (cfromknecht) [reply chainntnfs/btcdnotify/btcd.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 29 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 27 above)
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: removed in favor of panic on dereference

File: chainntnfs/bitcoindnotify/bitcoind.go [hunk 34]
```diff
@@ -254,38 +254,34 @@ out:
...
-					msg.TxID, msg.NumConfirmations)
-
-				currentHeight := uint32(b.bestBlock.Height)
-
-				// Look up whether the transaction is already
-				// included in the active chain. We'll do this
-				// in a goroutine to prevent blocking
-				// potentially long rescans.
+			case *chainntnfs.HistoricalConfDispatch:
+				// Look up whether the transaction is already included in the active
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/bitcoindnotify/bitcoind.go]: fixed

File: chainntnfs/txconfnotifier.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: fixed

File: chainntnfs/txconfnotifier.go [hunk 35]
```diff
@@ -299,13 +484,41 @@ func (tcn *TxConfNotifier) ConnectTip(blockHash *chainhash.Hash,
...
+
+		confSet.rescanStatus = rescanComplete
+		confSet.details = details
+		for _, ntfn := range confSet.ntfns {
+			// In the event that this notification was aware that
+			// the transaction was reorged out of the chain, we'll
+			// consume the reorg notification if it hasn't been done
+			// yet already.
+			select {
+			case <-ntfn.Event.NegativeConf:
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply chainntnfs/txconfnotifier.go]: `NegativeConf` is only buffered w/ size 1

File: chainntnfs/neutrinonotify/neutrino.go (hunk 32 above)
Comment (halseth) [reply chainntnfs/neutrinonotify/neutrino.go]: I was thinking that the code staring below `// If we can't fully dispatch` maybe could be removed now, but it looks like it must be there to update the scan filter with the address.

File: lnd_test.go [hunk 36]
```diff
@@ -7657,20 +7657,16 @@ func testGraphTopologyNotifications(net *lntest.NetworkHarness, t *harnessTest)
 	// The channel opening above should have triggered a few notifications
 	// sent to the notification client. We'll expect two channel updates,
 	// and two node announcements.
-	const numExpectedUpdates = 4
-	for i := 0; i < numExpectedUpdates; i++ {
+	var numChannelUpds int
```
Comment (Roasbeef) [inline lnd_test.go]: Reminder that this diff should go away after rebase, since it was included in: da56aa3528ae95398060de99101c4b7b5d2ae5c3

**Comment by cfromknecht** [comment]
Rebased
